pallet/dex/src/lib.rs

```diff
@@ -0,0 +1,525 @@
+		_,
+		Blake2_128Concat,
+		T::AccountId, //address
+		Blake2_128Concat,
+		u32, //asset id
+		TokenInfo,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn token_infoes)]
+	pub type TokenInfoes<T: Config> =
+		StorageValue<_, BoundedVec<u32, ConstU32<{ u32::MAX }>>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn token_index)]
+	pub type TokenIndex<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		u32,
+		u64, //token index
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn next_order_index)]
+	pub(super) type NextOrderIndex<T: Config> = StorageValue<_, u64, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn orders)]
+	pub type Orders<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		u64, //order index
+		OrderOf<T>,
+		OptionQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn next_pair_order_index)]
+	pub(super) type NextPairOrderIndex<T: Config> = StorageValue<_, u64, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn pair_orders)]
+	pub type PairOrders<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		(u32, u32),
+		BoundedVec<u64, ConstU32<{ u32::MAX }>>,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn user_orders)]
+	pub type UserOrders<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId, //address
+		Blake2_128Concat,
+		u64, //order index,
+		(),
+		ValueQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		SubmitProcessedReceipts {
+			block_number: u64,
+		},
+		OrderCreated {
+			order_index: u64,
+			order: OrderOf<T>,
+		},
+		OrderTaken {
+			account: T::AccountId,
+			order_index: u64,
+			order: OrderOf<T>,
+		},
+		OrderCanceled {
+			order_index: u64,
+		},
+		Deposited {
+			asset_id: u32,
+			amount: u128,
+		},
+
+		Withdrawed {
+			asset_id: u32,
+			amount: u128,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		OrderIndexOverflow,
+		InvalidOrderIndex,
+		InsufficientBalance,
+		NotOwner,
+		AssetIdNotInTokenIndex,
+		AssetIdNotInTokenInfoes,
+		TokenBalanceOverflow,
+		WithdrawBalanceMustKeepOrderSellAmount,
+		UserAssetNotExist,
+		PairOrderNotFound,
+		PairAssetIdMustNotEqual,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight({0})]
+		#[pallet::call_index(0)]
+		pub fn deposit(
+			origin: OriginFor<T>,
+			asset_id: u32,
+			amount: u128,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				TokenIndex::<T>::contains_key(asset_id),
+				Error::<T>::AssetIdNotInTokenIndex
+			);
+
+			<T::Fungibles as Mutate<T::AccountId>>::transfer(
+				asset_id,
+				&who,
+				&Self::account_id(),
+				amount,
+				Preservation::Expendable,
+			)?;
+
+			let mut info = TokenInfo::default();
+			if UserTokenInfoes::<T>::contains_key(who.clone(), asset_id) {
```

Comment:
I believe you can write lika that:

UserTokenInfoes::<T>::try_mutate(who.clone(), |info| {
   ...
 }.map_err(|_| UserTokenInfoes::insert...)